### DISCUSSION TITLE
Can I add an external go package for use in hkknx script?

### DISCUSSION THREAD

**Discussion by pfeifix** [brutella/hkknx-public#315] [General]
Ich suche aktuell nach einer Möglichkeit, Werte von go-Variablen auf dem System zu persistieren.

Ich habe ein bisschen gegoogelt und z. B. dieses package gefunden, was eine ganz geschickte API anzubieten scheint:

https://github.com/rapidloop/skv

Jetzt ist die Frage: Lässt sich so ein Paket in hkknx verwenden? Ich habe mal auf die Schnelle 

svk = import("github.com/rapidloop/svk")

im Skript ausprobiert, aber das scheint nicht zu funktionieren. Man müsste wahrscheinlich erst mal das Paket lokal installieren, aber bevor ich das mache, möchte ich Dich, @brutella erst einmal fragen, ob das überhaupt so funktionieren kann?

**Comment by brutella** [comment]
Bei, Skripte können nicht beliebige Packages importieren. Die Liste der verfügbaren Packages ist auf [der Webseite](https://hochgatterer.me/de/hkknx/scripts/packages/) einsehbar.

Die Möglichkeit der Speicherung von Daten besteht mit dem [os](https://hochgatterer.me/hkknx/scripts/pkgs/os/) Package. Mit der Method [os.WriteFile](https://hochgatterer.me/hkknx/scripts/pkgs/os/#WriteFile) können Daten in eine Datei gespeichert werden. Ein möglicher Dateipfad wäre `filepath.Join(os.TempDir(), "file.txt")` – das [path/filepath](https://hochgatterer.me/hkknx/scripts/pkgs/path/filepath/) Package steht auch zur Verfügung.

Hoffe damit kannst du was anfangen.

**Reply by brutella** [reply]
@pfeifix Konntest du mit Hilfe meiner Anleitung die erforderlichen Daten speichern?

**Reply by pfeifix** [reply]
Danke, ja das hilft erst mal zur Einordnung.

Allerdings finde ich mit den Weg über das os-Package mit den rohen Dateizugriffen etwas umständlich (die Skripte werden recht lang), wenn es nur darum geht, ein einfaches Key-Value-Dictionary zu halten. Eigentlich wäre für mich eine Programmierlösung, mit der man wiederverwendbare Funktionen schreiben und über die Skripte aufrufen könnte, am liebsten (dann wäre es egal, wie lang der Code wird, weil man ihn ja nur an einer Stelle hat), sehe aber, dass das eine nicht schnell umzusetzende Erweiterung von hkknx darstellen würde.

Gibt es irgendein mitgeliefertes Package, mit dem man Schlüssel-Wert-Paare mit ein paar wenigen Zeilen Code auf die Platte schreiben und wieder lesen könnte?

**Reply by brutella** [reply]
> Gibt es irgendein mitgeliefertes Package, mit dem man Schlüssel-Wert-Paare mit ein paar wenigen Zeilen Code auf die Platte schreiben und wieder lesen könnte?

Nein, das gibt es (noch) nicht. Was sollte deiner Meinung nach so ein Package können?

Werte speichern und auslesen?

```go
var hkknx = import("hkknx")

hkknx.Set("username", "matthias")
var username = hkknx.Get("username") // username == "matthias"
```

- Sollen diese Daten auch für andere Skripte zugänglich sein?
- Wie soll man diese Daten bearbeiten oder löschen können? (Irgendwo muss man sehen können, welche Daten gespeichert sind und wieviel Speicherplatz diese benötigen.) 
- Soll es eine maximale Datengröße geben?

**Reply by pfeifix** [reply]
Die Schlüssel-Wert-Paare sollten über alle Skripte hinweg verfügbar sein. Wie die Datenhaltung im Hintergrund organisiert ist, wäre für mich als Nutzer erst einmal nebensächlich. Natürlich sollte man irgendwo sehen, was gespeichert ist und die Werte bzw. das ganze Dictionary bei Bedarf löschen können. Ob dies per API oder die GUI erfolgt, wäre für mich auch erst einmal zweitrangig, wobei eine Lösung über die Oberfläche natürlich einen gewissen Charme hätte.
